src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java

```diff
@@ -139,6 +143,19 @@ public SCM build(SCMHead head, @CheckForNull SCMRevision revision) {
         return gitSCMSource.build(head, revision);
     }
 
+    @Override
+    public void afterSave() {
+        super.afterSave();
+        if (!webhookRegistered && isValid()) {
+            SCMSourceOwner owner = getOwner();
+            if (owner instanceof ComputedFolder) {
```

Comment:
Why not `WorkflowMultiBranchProject`? That's the only one our source supports.

Reply:
No particular reason other than the general preference to use the most abstract type in a type hierarchy that provides the desired _behaviour_ and/or _attribute_, as opposed to being unnecessarily explicit and restrictive.
Just as a principle, even though we _know_ that in this case we only support `WorkflowMultiBranchProject`, there is no static and contractual boundary in the code on what type can/will be passed in here, so (IMO) restricting it to `WorkflowMultiBranchProject` will make the code unnecessarily inflexible, in case a different subtype of `ComputedFolder` is ever passed in in the future for any reason.

Also, even if the `WorkflowMultiBranchProject` is to be enforced as the `SCMSourceOwner` type here, it is not the _responsibility_ of this `if` block to do so. This `if` specifically checks that the `SCMSourceOwner` of the project _can contain_ `trigger`s, before getting those triggers and registering webhooks with them, etc.
In an ideal world (i.e. not in Jenkins), `ComputedFolder` would instead be an interface (e.g. `SCMOwnersThatContainTriggers`) that marks the responsibility/behaviour of containing triggers, and we would use the interface for referencing the `owner` object everywhere here, including when passing it to `getTriggers(ComputedFolder<?> owner)`.

From [Effective Java Item 52](http://www.thefinestartist.com/effective-java/52):
> If appropriate interface types exist, then parameters, return values, variables, and fields should all be declared using interface types.

My interpretation of that is, if an interface doesn't exist, the abstract super-type is the second best thing.